perf(transformer/jsx): use `memchr` for parsing JSX pragma comments

overlookmotel opened the pull request:
Use `memchr` for finding `@` when parsing JSX pragmas from comments.

This wins back most (but not all) of the perf loss of #10983 on `antd.js` benchmark, and preserves the perf gain of #10983 on `cal.com.tsx` benchmark.

Interestingly, using `memchr` to search just for `@` and then checking next 3 bytes are `jsx` separately is measurably faster than using `memchr::memmem::Finder` to search for `@jsx`.

Boshen commented:
### Merge activity

* **May 14, 12:46 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(transformer/jsx): use `memchr` for parsing JSX pragma comments (#11001)

Use `memchr` for finding `@` when parsing JSX pragmas from comments.

This wins back most (but not all) of the perf loss of #10983 on `antd.js` benchmark, and preserves the perf gain of #10983 on `cal.com.tsx` benchmark.

Interestingly, using `memchr` to search just for `@` and then checking next 3 bytes are `jsx` separately is measurably faster than using `memchr::memmem::Finder` to search for `@jsx`.

camc314 commented on crates/oxc_transformer/src/jsx/comments.rs:
```diff
@@ -96,20 +98,26 @@ enum PragmaType {
+        // Search for `@`.
+        // Note: Using `memchr::memmem::Finder` to search for `@jsx` is slower than only using `memchr`
+        // to find `@` characters, and then checking if `@` is followed by `jsx` separately.
+        let at_sign_index = memchr(b'@', comment_str.as_bytes())?;
```
it might also be faster to use memchr_iter rather than a loop and having the searcher re-compiled again. (it also might make no difference depending on how the compiler is optimizing it)

overlookmotel replied:
I don't think `memchr` (unlike `memchr::memmem::Finder`) compiles a searcher type - it's a straight function call. If it's inlined, the "needle" (`@`) should be statically known to compiler, which might allow it to generate more optimal code than a more complex type.

I've taken some shortcuts here - there are some bounds checks which could be eliminated, and other stuff that could be a little tighter. But my assumption was that almost all the cost is in searching for `@`, so everything after that was not worthwhile optimizing.

But still, yes, `memchr_iter` might be faster, especially if the searcher was created outside of `find_jsx_pragma` and passed in so it can be re-used. Want to try it?
